pkg/volume/cinder/cinder.go

```diff
@@ -0,0 +1,289 @@
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+// This is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&cinderPlugin{nil}}
+}
+
+type cinderPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &cinderPlugin{}
+
+const (
+	cinderVolumePluginName = "kubernetes.io/cinder"
+)
+
+func (plugin *cinderPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *cinderPlugin) Name() string {
+	return cinderVolumePluginName
+}
+
+func (plugin *cinderPlugin) CanSupport(spec *volume.Spec) bool {
+	return spec.PersistentVolumeSource.Cinder != nil || spec.VolumeSource.Cinder != nil
+}
+
+func (plugin *cinderPlugin) GetAccessModes() []api.PersistentVolumeAccessMode {
+	return []api.PersistentVolumeAccessMode{
+		api.ReadWriteOnce,
+		api.ReadOnlyMany,
+	}
+}
+
+func (plugin *cinderPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, _ volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, pod.UID, &CinderDiskUtil{}, mounter)
+}
+
+func (plugin *cinderPlugin) newBuilderInternal(spec *volume.Spec, podUID types.UID, manager cdManager, mounter mount.Interface) (volume.Builder, error) {
+
+	var cinder *api.CinderVolumeSource
+	if spec.VolumeSource.Cinder != nil {
+		cinder = spec.VolumeSource.Cinder
+	} else {
+		cinder = spec.PersistentVolumeSource.Cinder
+	}
+
+	pdName := cinder.VolID
+	fsType := cinder.FSType
+	readOnly := cinder.ReadOnly
+
+	return &cinderVolumeBuilder{
+		cinderVolume: &cinderVolume{
+			podUID:  podUID,
+			volName: spec.Name,
+			pdName:  pdName,
+			mounter: mounter,
+			manager: manager,
+			plugin:  plugin,
+		},
+		fsType:      fsType,
+		readOnly:    readOnly,
+		blockDeviceMounter: &cinderSafeFormatAndMount{mounter, exec.New()}}, nil
+}
+
+func (plugin *cinderPlugin) NewCleaner(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, &CinderDiskUtil{}, mounter)
+}
+
+func (plugin *cinderPlugin) newCleanerInternal(volName string, podUID types.UID, manager cdManager, mounter mount.Interface) (volume.Cleaner, error) {
+
+	return &cinderVolumeCleaner{
+		&cinderVolume{
+			podUID:  podUID,
+			volName: volName,
+			manager: manager,
+			mounter: mounter,
+			plugin:  plugin,
+		}}, nil
+}
+
+// Abstract interface to PD operations.
+type cdManager interface {
+	// Attaches the disk to the kubelet's host machine.
+	AttachDisk(builder *cinderVolumeBuilder, globalPDPath string) error
+	// Detaches the disk from the kubelet's host machine.
+	DetachDisk(cleaner *cinderVolumeCleaner) error
+}
+
+var _ volume.Builder = &cinderVolumeBuilder{}
+
+type cinderVolumeBuilder struct {
+	*cinderVolume
+
+	fsType string
+
+	readOnly bool
+
+    blockDeviceMounter mount.Interface
```

Comment:
Surprised @thockin didn't nail you on blank lines :)

You can remove those empty spaces between struct vars.